prow/cmd/deck/job_history.go

```diff
@@ -240,16 +256,39 @@ func (bucket gcsBucket) listBuildIDs(root string) ([]int64, error) {
 	return ids, nil
 }
 
-func parseJobHistURL(url *url.URL) (bucketName, root string, buildID int64, err error) {
+// parseJobHistURL parses the job History URL
+// example urls:
+// * new format: https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-capi?buildId=1245584383100850177
+// * old format: https://prow.k8s.io/job-history/kubernetes-jenkins/pr-logs/directory/pull-capi?buildId=1245584383100850177
+// Newly generated URLs will include the storageProvider. We still support old URLs so they don't break.
+// For old URLs we assume that the storageProvider is `gs`.
+// examples return values:
+// * storageProvider: s3, gs
+// * bucketName: kubernetes-jenkins
+// * root: pr-logs/directory/pull-capi
+// * buildID: 1245584383100850177
+func parseJobHistURL(url *url.URL) (storageProvider, bucketName, root string, buildID int64, err error) {
 	buildID = emptyID
 	p := strings.TrimPrefix(url.Path, "/job-history/")
-	s := strings.SplitN(p, "/", 2)
-	if len(s) < 2 {
-		err = fmt.Errorf("invalid path (expected /job-history/<gcs-path>): %v", url.Path)
+	// examples for p:
+	// * new format: gs/prow-artifacts/pr-logs/directory/bazel-build-test
+	// * old format: kubernetes-jenkins/pr-logs/directory/pull-cluster-api-provider-openstack-test
+
+	// inject gs/ if old format is used
+	if !strings.HasPrefix(p, providers.GS+"/") && !strings.HasPrefix(p, providers.S3+"/") {
```

Comment:
Instead of doing this can we make the URL:

`gs://prow-aritfacts/pr-logs/directory`

rather than just `gs/prow-artifacts/pr-logs/directory`

and if there is no scheme in the path then add gs://`?

https://golang.org/pkg/net/url/#URL
```golang
package main

import (
	"fmt"
	"log"
	"net/url"
)

func main() {
	u, err := url.Parse("http://bing.com/search?q=dotnet")
	if err != nil {
		log.Fatal(err)
	}
	u.Scheme = "https"
	u.Host = "google.com"
	q := u.Query()
	q.Set("q", "golang")
	u.RawQuery = q.Encode()
	u.Path = "gs://hello/world"
	fmt.Println(u) // https://google.com/gs://hello/world?q=golang
}
```

Reply:
I'm not sure if I got it right. Do you mean changing the external URL (the one used by the browser) to something like: `https://prow.k8s.io/job-history/gs://kubernetes-jenkins/pr-logs/directory/pull-capi?buildId=1245584383100850177`

If we want to do that we have to be really careful as every path.Join we're doing anywhere will break it by changing "gs://" to "gs:/"

Reply:
I mostly just don't like that we're special casing providers here and trying to think of an alternative.

How about if we whitelist the set of buckets that we will redirect? AKA
`--default-logs-provider=gs --migrate-bucket=kubernetes-jenkins --migrate-bucket=other-bucket` (`flagutils.Strings` will support flag repetition)

And so then if `p` starts with `kubernetes-jenkins/` or `other-bucket/` we'll replace it with `gs/kubernetes-jenkins` as appropriate.

Having a temporary `--require-provider=gs` would also work (if p doesn't start with gs then insert this at the front, and log that this support will go away by X and they need to set --default-logs-provider=gs and --migrate-bucket)

Reply:
If possible, it would be nice to avoid additional flags for this.

We have to differentiate between:
*  `<storage-provider>/<bucket>/<path>`
*  `<bucket>/<path>`

There are two ways of doing this:
* detect that the string starts with a valid storage provider
* detect that the string starts with a valid bucket (which could be a lot of different ones)

I guess we still have to handle the old format for a while as we don't want to risk broken links.
The main problem of the current solution is (I guess) that we have to enumerate all valid `storageProviders` and if we add a new one it also has to be added here.

We are already exposing the constants `providers.S3` and `providers.GS` in the `providers` package. What do you think about adding a function like `providers.HasStorageProviderPrefix` to the `providers` package?

I think that's considering the tradeoffs the best solution.

Reply:
That sounds perfectly reasonable, yes.

Can you either make that change or else squash your commits (or both)? Then LGTM from me.

Reply:
Done. I pushed the commit invidually. I'll squash it all now